File: packages/sphere_view/android/src/main/kotlin/com/asite/sphereview/TimestampMapper.kt
package com.asite.sphereview

import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraMetadata
import android.os.SystemClock

/**
 * Puts `SENSOR_TIMESTAMP` on the same clock as `SensorEvent.timestamp`.
 *
 * Phase 06 §2.5 calls this the highest-consequence detail in the phase, and the
 * reason is arithmetic rather than pedantry: a 10–50 ms pose/frame mismatch is
 * 0.6–3° of rotation error at a realistic 60°/s pan. That is larger than
 * everything Phase 03 works to remove, and it does not present as a clock bug —
 * it presents as a stitcher that will not close a seam.
 *
 * There are exactly two cases, named by `SENSOR_INFO_TIMESTAMP_SOURCE`:
 *
 *  - **REALTIME** — the camera stamps frames with `elapsedRealtimeNanos()`,
 *    which is also the base of `SensorEvent.timestamp`. The two *are* one
 *    clock. The offset is zero, exactly, and estimating it would only add
 *    noise.
 *  - **UNKNOWN** — the camera stamps with `System.nanoTime()`, which is not the
 *    sensor base. The offset is real, is device-specific, and has to be
 *    measured.
 *
 * The measurement is the standard interleaved read: bracket a `nanoTime()`
 * reading between two `elapsedRealtimeNanos()` readings and keep the sample
 * whose bracket was tightest. §2.5 describes this as "sampling both clocks in a
 * tight loop and taking the minimum observed difference"; taking the minimum of
 * a raw difference works only because the read latency is one-sided, whereas
 * keeping the narrowest bracket bounds the error explicitly — and the bound is
 * what gets reported, because the exit criterion is stated in microseconds and
 * an unqualified number could not be checked against it.
 */
class TimestampMapper private constructor(
    val base: PlatformTimestampBase,
    /** Nanoseconds to add to a raw camera timestamp to reach the motion clock. */
    val offsetNs: Long,
    /** Half the narrowest read bracket achieved, in nanoseconds. */
    val uncertaintyNs: Long,
    val note: String,
) {

    companion object {
        /**
         * How many interleaved reads to take. 128 costs well under a
         * millisecond and makes it very likely at least one bracket lands
         * inside a scheduling quantum.
         */
        private const val SAMPLES = 128

        fun forCamera(characteristics: CameraCharacteristics): TimestampMapper {
            val source = characteristics.get(CameraCharacteristics.SENSOR_INFO_TIMESTAMP_SOURCE)
            return when (source) {
                CameraMetadata.SENSOR_INFO_TIMESTAMP_SOURCE_REALTIME ->
                    TimestampMapper(
                        base = PlatformTimestampBase.ANDROID_REALTIME,
                        offsetNs = 0L,
                        uncertaintyNs = 0L,
                        note =
                            "SENSOR_INFO_TIMESTAMP_SOURCE is REALTIME, so SENSOR_TIMESTAMP is " +
                                "already on elapsedRealtimeNanos() — the same base as " +
                                "SensorEvent.timestamp. No conversion, no estimate, no error.",
                    )

                else -> {
                    val m = measure()
                    TimestampMapper(
                        base = PlatformTimestampBase.ANDROID_MONOTONIC_UNKNOWN,
                        offsetNs = m.offsetNs,
                        uncertaintyNs = m.uncertaintyNs,
                        note =
                            "SENSOR_INFO_TIMESTAMP_SOURCE is " +
                                (if (source == null) "unset" else "UNKNOWN") +
                                ", so SENSOR_TIMESTAMP is on System.nanoTime() and NOT " +
                                "comparable with SensorEvent.timestamp. Offset measured over " +
                                "$SAMPLES interleaved reads: ${m.offsetNs / 1000} µs, narrowest " +
                                "read bracket ${m.windowNs / 1000} µs, spread across samples " +
                                "${m.spreadNs / 1000} µs.",
                    )
                }
            }
        }

        /** Re-measures now, for the §6 stability check. */
        fun measure(): Estimate {
            var bestWindow = Long.MAX_VALUE
            var bestOffset = 0L
            var minOffset = Long.MAX_VALUE
            var maxOffset = Long.MIN_VALUE
            for (i in 0 until SAMPLES) {
                val before = SystemClock.elapsedRealtimeNanos()
                val mono = System.nanoTime()
                val after = SystemClock.elapsedRealtimeNanos()
                val window = after - before
                // Midpoint of the bracket is the best estimate of where the
                // realtime clock stood when nanoTime() was read.
                val offset = (before + after) / 2 - mono
                if (offset < minOffset) minOffset = offset
                if (offset > maxOffset) maxOffset = offset
                if (window < bestWindow) {
                    bestWindow = window
                    bestOffset = offset
                }
            }
            return Estimate(
                offsetNs = bestOffset,
                uncertaintyNs = bestWindow / 2,
                windowNs = bestWindow,
                spreadNs = if (maxOffset >= minOffset) maxOffset - minOffset else 0L,
            )
        }
    }

    /** One offset measurement and the two ways of describing its error. */
    data class Estimate(
        val offsetNs: Long,
        val uncertaintyNs: Long,
        val windowNs: Long,
        val spreadNs: Long,
    )

    /** Converts a raw camera `SENSOR_TIMESTAMP` to motion-clock microseconds. */
    fun toMotionClockUs(sensorTimestampNs: Long): Long = (sensorTimestampNs + offsetNs) / 1000L

    /** The motion clock's current reading, in microseconds. */
    fun motionClockNowUs(): Long = SystemClock.elapsedRealtimeNanos() / 1000L

    /** The camera clock's current reading, in microseconds. */
    fun cameraClockNowUs(): Long =
        when (base) {
            PlatformTimestampBase.ANDROID_REALTIME -> SystemClock.elapsedRealtimeNanos() / 1000L
            else -> System.nanoTime() / 1000L
        }

    fun toInfo(): ClockSyncInfo =
        ClockSyncInfo(
            base = base,
            offsetUs = offsetNs / 1000L,
            uncertaintyUs = uncertaintyNs / 1000L,
            note = note,
        )

    /**
     * A fresh sample for the §6 stability test. On REALTIME this is trivially
     * zero every time — which is the correct answer and worth returning rather
     * than skipping, because "the offset never moved" is exactly what the test
     * is checking.
     */
    fun sample(): ClockOffsetSample {
        val camera = cameraClockNowUs()
        val motion = motionClockNowUs()
        return when (base) {
            PlatformTimestampBase.ANDROID_REALTIME ->
                ClockOffsetSample(
                    cameraClockUs = camera,
                    motionClockUs = motion,
                    offsetUs = 0L,
                    uncertaintyUs = 0L,
                )

            else -> {
                val m = measure()
                ClockOffsetSample(
                    cameraClockUs = camera,
                    motionClockUs = motion,
                    offsetUs = m.offsetNs / 1000L,
                    uncertaintyUs = m.uncertaintyNs / 1000L,
                )
            }
        }
    }
}

File: packages/sphere_view/ios/Classes/TimestampMapper.swift
import CoreMedia
import Foundation
import QuartzCore

/// Puts `CMSampleBuffer` presentation timestamps on the same clock as
/// `CMDeviceMotion.timestamp`.
///
/// §3.5 states the problem in one line — sample buffers are on
/// `CMClockGetHostTimeClock` (`mach_absolute_time`) while `CMDeviceMotion`
/// reports `systemUptime` — and §2.5 states the consequence of getting it
/// wrong: a 10–50 ms pose/frame mismatch is 0.6–3° of rotation error at a
/// realistic 60°/s pan, which is larger than everything Phase 03 works to
/// remove and which presents as a stitcher bug rather than as a clock bug.
///
/// On current iOS the two clocks are the same clock underneath, so the offset
/// is a handful of microseconds of read latency and nothing more. **That is
/// measured here rather than assumed**, for two reasons. It is an undocumented
/// implementation detail, not a contract — Apple documents `systemUptime` as
/// "the amount of time the system has been awake since the last time it was
/// restarted" and documents the host clock separately, and nothing promises
/// they share an epoch on some future device or on a platform variant. And the
/// exit criterion is stated as ±2 ms of stability over 60 s, which is a claim
/// about a measurement; asserting equality would leave nothing to check.
final class TimestampMapper {

    /// How many interleaved reads to take. Cheap — the whole loop is well under
    /// a millisecond — and enough that one bracket usually lands inside a
    /// scheduling quantum.
    private static let sampleCount = 128

    /// Microseconds to add to a host-clock timestamp to reach the motion clock.
    private(set) var offsetUs: Int64

    /// Half the narrowest read bracket achieved, in microseconds.
    private(set) var uncertaintyUs: Int64

    private(set) var note: String

    init() {
        let estimate = TimestampMapper.measure()
        offsetUs = estimate.offsetUs
        uncertaintyUs = estimate.uncertaintyUs
        note =
            "CMSampleBuffer presentation timestamps are on CMClockGetHostTimeClock; "
            + "CMDeviceMotion.timestamp is on ProcessInfo.systemUptime. Offset measured over "
            + "\(TimestampMapper.sampleCount) interleaved reads: \(estimate.offsetUs) µs, "
            + "narrowest read bracket \(estimate.windowUs) µs, spread across samples "
            + "\(estimate.spreadUs) µs. A near-zero offset is the expected result on current "
            + "iOS — the two are the same underlying clock — but it is measured, not assumed."
    }

    struct Estimate {
        let offsetUs: Int64
        let uncertaintyUs: Int64
        let windowUs: Int64
        let spreadUs: Int64
    }

    /// Brackets a `systemUptime` read between two host-clock reads and keeps the
    /// sample whose bracket was tightest, which bounds the error explicitly
    /// instead of hoping the two reads were adjacent.
    static func measure() -> Estimate {
        var bestWindow = Double.greatestFiniteMagnitude
        var bestOffset = 0.0
        var minOffset = Double.greatestFiniteMagnitude
        var maxOffset = -Double.greatestFiniteMagnitude

        for _ in 0..<sampleCount {
            let before = hostTimeSeconds()
            let motion = ProcessInfo.processInfo.systemUptime
            let after = hostTimeSeconds()
            let window = after - before
            let offset = (before + after) / 2 - motion
            minOffset = min(minOffset, offset)
            maxOffset = max(maxOffset, offset)
            if window < bestWindow {
                bestWindow = window
                bestOffset = offset
            }
        }

        // `offset` above is host − motion, i.e. how far ahead the host clock
        // reads. To move a host timestamp onto the motion clock, subtract it.
        return Estimate(
            offsetUs: Int64((-bestOffset * 1e6).rounded()),
            uncertaintyUs: Int64((bestWindow / 2 * 1e6).rounded()),
            windowUs: Int64((bestWindow * 1e6).rounded()),
            spreadUs: Int64(((maxOffset - minOffset) * 1e6).rounded())
        )
    }

    /// The host clock's current reading in seconds, via the same clock the
    /// capture pipeline stamps buffers with.
    static func hostTimeSeconds() -> Double {
        CMTimeGetSeconds(CMClockGetTime(CMClockGetHostTimeClock()))
    }

    /// Converts a sample-buffer presentation timestamp to motion-clock
    /// microseconds.
    func toMotionClockUs(_ time: CMTime) -> Int64 {
        guard time.isValid && !time.seconds.isNaN else { return 0 }
        return Int64((time.seconds * 1e6).rounded()) + offsetUs
    }

    /// Converts a host-clock instant in seconds, for the callbacks that report
    /// one rather than a `CMTime`.
    func toMotionClockUs(hostSeconds: Double) -> Int64 {
        Int64((hostSeconds * 1e6).rounded()) + offsetUs
    }

    func toInfo() -> ClockSyncInfo {
        ClockSyncInfo(
            base: .iosHostTime,
            offsetUs: offsetUs,
            uncertaintyUs: uncertaintyUs,
            note: note
        )
    }

    /// A fresh sample for the §6 stability test.
    func sample() -> ClockOffsetSample {
        let estimate = TimestampMapper.measure()
        return ClockOffsetSample(
            cameraClockUs: Int64((TimestampMapper.hostTimeSeconds() * 1e6).rounded()),
            motionClockUs: Int64((ProcessInfo.processInfo.systemUptime * 1e6).rounded()),
            offsetUs: estimate.offsetUs,
            uncertaintyUs: estimate.uncertaintyUs
        )
    }
}
